File: android-native/core/network/src/main/kotlin/today/mindlog/todo/core/network/api/TodoApi.kt
package today.mindlog.todo.core.network.api

import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query
import today.mindlog.todo.core.network.model.Filter
import today.mindlog.todo.core.network.model.FilterCreateRequest
import today.mindlog.todo.core.network.model.FilterUpdateRequest
import today.mindlog.todo.core.network.model.Label
import today.mindlog.todo.core.network.model.LabelCreateRequest
import today.mindlog.todo.core.network.model.LabelUpdateRequest
import today.mindlog.todo.core.network.model.Project
import today.mindlog.todo.core.network.model.ProjectCreateRequest
import today.mindlog.todo.core.network.model.ProjectUpdateRequest
import today.mindlog.todo.core.network.model.Section
import today.mindlog.todo.core.network.model.SectionCreateRequest
import today.mindlog.todo.core.network.model.SectionUpdateRequest
import today.mindlog.todo.core.network.model.AskResult
import today.mindlog.todo.core.network.model.Task
import today.mindlog.todo.core.network.model.TaskAskRequest
import today.mindlog.todo.core.network.model.TaskSearchHit
import today.mindlog.todo.core.network.model.TaskSearchRequest
import today.mindlog.todo.core.network.model.TaskCreateRequest
import today.mindlog.todo.core.network.model.TaskQuickAddRequest
import today.mindlog.todo.core.network.model.TaskUpdateRequest
import today.mindlog.todo.core.network.model.User

/**
 * The authenticated surface. Hand-written on purpose — it is thirty lines,
 * while generating it would drag in openapi-generator's own ApiClient and auth
 * handling, neither of which fits the client this module builds.
 *
 * Paths are relative; see the note on [AuthApi].
 */
interface TodoApi {

    @GET("api/v1/me")
    suspend fun me(): User

    // --- tasks ---

    /**
     * `limit` defaults to 50 server-side and caps at 200. Milestone 1 asks for
     * the maximum and does not paginate; the day a user has more than two
     * hundred open tasks, that becomes a real gap rather than a hidden one.
     */
    @GET("api/v1/tasks")
    suspend fun listTasks(
        @Query("completed") completed: Boolean? = null,
        @Query("root") root: Boolean? = null,
        @Query("projectId") projectId: String? = null,
        @Query("labelId") labelId: String? = null,
        /** ISO instant. `Today` is expressed as "due before tomorrow", so it includes overdue. */
        @Query("dueBefore") dueBefore: String? = null,
        @Query("limit") limit: Int = 200,
    ): List<Task>

    @POST("api/v1/tasks")
    suspend fun createTask(@Body body: TaskCreateRequest): Task

    /**
     * One line of natural language, parsed server-side: dates, `#project`,
     * `@label`, `p1`–`p4`. Using this instead of a form is what keeps the
     * parsing rules in exactly one place.
     */
    @POST("api/v1/tasks/quickadd")
    suspend fun quickAdd(@Body body: TaskQuickAddRequest): Task

    @PATCH("api/v1/tasks/{id}")
    suspend fun updateTask(@Path("id") id: String, @Body body: TaskUpdateRequest): Task

    @DELETE("api/v1/tasks/{id}")
    suspend fun deleteTask(@Path("id") id: String)

    /** Recherche sémantique sur les tâches ; le score est porté par le hit. */
    @POST("api/v1/tasks/search")
    suspend fun searchTasks(@Body body: TaskSearchRequest): List<TaskSearchHit>

    /**
     * Question en langue naturelle. La réponse cite les tâches qui l'ont
     * fondée : elles sont rendues telles quelles, pas résumées, pour que
     * l'utilisateur puisse vérifier ce sur quoi le modèle s'appuie.
     */
    @POST("api/v1/tasks/ask")
    suspend fun ask(@Body body: TaskAskRequest): AskResult

    // --- projects ---

    @GET("api/v1/projects")
    suspend fun listProjects(
        @Query("includeArchived") includeArchived: Boolean? = null,
    ): List<Project>

    @GET("api/v1/projects/{id}")
    suspend fun getProject(@Path("id") id: String): Project

    @POST("api/v1/projects")
    suspend fun createProject(@Body body: ProjectCreateRequest): Project

    @PATCH("api/v1/projects/{id}")
    suspend fun updateProject(@Path("id") id: String, @Body body: ProjectUpdateRequest): Project

    @DELETE("api/v1/projects/{id}")
    suspend fun deleteProject(@Path("id") id: String)

    // --- sections ---

    /**
     * `projectId` is required by the server: a section only exists inside a
     * project, and listing them all at once is not offered.
     */
    @GET("api/v1/sections")
    suspend fun listSections(@Query("projectId") projectId: String): List<Section>

    @POST("api/v1/sections")
    suspend fun createSection(@Body body: SectionCreateRequest): Section

    @PATCH("api/v1/sections/{id}")
    suspend fun updateSection(@Path("id") id: String, @Body body: SectionUpdateRequest): Section

    @DELETE("api/v1/sections/{id}")
    suspend fun deleteSection(@Path("id") id: String)

    // --- labels ---

    @GET("api/v1/labels")
    suspend fun listLabels(): List<Label>

    @POST("api/v1/labels")
    suspend fun createLabel(@Body body: LabelCreateRequest): Label

    @PATCH("api/v1/labels/{id}")
    suspend fun updateLabel(@Path("id") id: String, @Body body: LabelUpdateRequest): Label

    @DELETE("api/v1/labels/{id}")
    suspend fun deleteLabel(@Path("id") id: String)

    // --- filters (saved views) ---

    @GET("api/v1/filters")
    suspend fun listFilters(): List<Filter>

    @POST("api/v1/filters")
    suspend fun createFilter(@Body body: FilterCreateRequest): Filter

    @PATCH("api/v1/filters/{id}")
    suspend fun updateFilter(@Path("id") id: String, @Body body: FilterUpdateRequest): Filter

    @DELETE("api/v1/filters/{id}")
    suspend fun deleteFilter(@Path("id") id: String)

    /**
     * Runs the saved query server-side. The client never interprets the filter
     * syntax itself — that grammar lives in one place, on the server.
     */
    @GET("api/v1/filters/{id}/tasks")
    suspend fun runFilter(@Path("id") id: String): List<Task>
}

File: ios-native/Sources/CoreNetwork/TodoAPI.swift
import Foundation

/// The authenticated surface. Hand-written on purpose — it is a list of paths,
/// while generating it would drag in openapi-generator's own client, auth
/// handling and error types, none of which fit the stack this module builds.
/// It is `TodoTask` and its twenty fields that must never be retyped, and those
/// *are* generated.
///
/// Paths are relative; see the note on ``AuthAPI``.
public struct TodoAPI: Sendable {

    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    public func me() async throws(APIError) -> TodoUser {
        try await client.send(.init(.get, "api/v1/me"), as: TodoUser.self)
    }

    // MARK: - Tasks

    /// `limit` defaults to 50 server-side and caps at 200. Milestone 1 asks for
    /// the maximum and does not paginate; the day a user has more than two
    /// hundred open tasks, that becomes a real gap rather than a hidden one.
    ///
    /// - Parameter dueBefore: ISO instant. "Today" is expressed as "due before
    ///   tomorrow", so it includes overdue.
    public func listTasks(
        completed: Bool? = nil,
        root: Bool? = nil,
        projectId: String? = nil,
        labelId: String? = nil,
        dueBefore: String? = nil,
        limit: Int = 200
    ) async throws(APIError) -> [TodoTask] {
        try await client.send(
            .init(.get, "api/v1/tasks", query: .query([
                ("completed", completed.map(String.init)),
                ("root", root.map(String.init)),
                ("projectId", projectId),
                ("labelId", labelId),
                ("dueBefore", dueBefore),
                ("limit", String(limit)),
            ])),
            as: [TodoTask].self
        )
    }

    public func createTask(_ body: TodoTaskCreateRequest) async throws(APIError) -> TodoTask {
        try await client.send(
            .init(.post, "api/v1/tasks", body: JSONBody.encode(body)),
            as: TodoTask.self
        )
    }

    /// One line of natural language, parsed server-side: dates, `#project`,
    /// `@label`, `p1`–`p4`. Using this instead of a form is what keeps the
    /// parsing rules in exactly one place.
    public func quickAdd(_ body: TodoTaskQuickAddRequest) async throws(APIError) -> TodoTask {
        try await client.send(
            .init(.post, "api/v1/tasks/quickadd", body: JSONBody.encode(body)),
            as: TodoTask.self
        )
    }

    public func updateTask(
        _ id: String,
        _ body: TodoTaskUpdateRequest
    ) async throws(APIError) -> TodoTask {
        try await client.send(
            .init(.patch, "api/v1/tasks/\(id)", body: JSONBody.encode(body)),
            as: TodoTask.self
        )
    }

    public func deleteTask(_ id: String) async throws(APIError) {
        try await client.send(.init(.delete, "api/v1/tasks/\(id)"))
    }

    // MARK: - Projects

    public func listProjects(
        includeArchived: Bool? = nil
    ) async throws(APIError) -> [TodoProject] {
        try await client.send(
            .init(.get, "api/v1/projects", query: .query([
                ("includeArchived", includeArchived.map(String.init)),
            ])),
            as: [TodoProject].self
        )
    }

    public func getProject(_ id: String) async throws(APIError) -> TodoProject {
        try await client.send(.init(.get, "api/v1/projects/\(id)"), as: TodoProject.self)
    }

    public func createProject(
        _ body: TodoProjectCreateRequest
    ) async throws(APIError) -> TodoProject {
        try await client.send(
            .init(.post, "api/v1/projects", body: JSONBody.encode(body)),
            as: TodoProject.self
        )
    }

    public func updateProject(
        _ id: String,
        _ body: TodoProjectUpdateRequest
    ) async throws(APIError) -> TodoProject {
        try await client.send(
            .init(.patch, "api/v1/projects/\(id)", body: JSONBody.encode(body)),
            as: TodoProject.self
        )
    }

    public func deleteProject(_ id: String) async throws(APIError) {
        try await client.send(.init(.delete, "api/v1/projects/\(id)"))
    }

    // MARK: - Sections

    /// `projectId` is required by the server: a section only exists inside a
    /// project, and listing them all at once is not offered.
    public func listSections(projectId: String) async throws(APIError) -> [TodoSection] {
        try await client.send(
            .init(.get, "api/v1/sections", query: .query([("projectId", projectId)])),
            as: [TodoSection].self
        )
    }

    public func createSection(
        _ body: TodoSectionCreateRequest
    ) async throws(APIError) -> TodoSection {
        try await client.send(
            .init(.post, "api/v1/sections", body: JSONBody.encode(body)),
            as: TodoSection.self
        )
    }

    public func updateSection(
        _ id: String,
        _ body: TodoSectionUpdateRequest
    ) async throws(APIError) -> TodoSection {
        try await client.send(
            .init(.patch, "api/v1/sections/\(id)", body: JSONBody.encode(body)),
            as: TodoSection.self
        )
    }

    public func deleteSection(_ id: String) async throws(APIError) {
        try await client.send(.init(.delete, "api/v1/sections/\(id)"))
    }

    // MARK: - Labels

    public func listLabels() async throws(APIError) -> [TodoLabel] {
        try await client.send(.init(.get, "api/v1/labels"), as: [TodoLabel].self)
    }

    public func createLabel(_ body: TodoLabelCreateRequest) async throws(APIError) -> TodoLabel {
        try await client.send(
            .init(.post, "api/v1/labels", body: JSONBody.encode(body)),
            as: TodoLabel.self
        )
    }

    public func updateLabel(
        _ id: String,
        _ body: TodoLabelUpdateRequest
    ) async throws(APIError) -> TodoLabel {
        try await client.send(
            .init(.patch, "api/v1/labels/\(id)", body: JSONBody.encode(body)),
            as: TodoLabel.self
        )
    }

    public func deleteLabel(_ id: String) async throws(APIError) {
        try await client.send(.init(.delete, "api/v1/labels/\(id)"))
    }

    // MARK: - Filters (saved views)

    public func listFilters() async throws(APIError) -> [TodoFilter] {
        try await client.send(.init(.get, "api/v1/filters"), as: [TodoFilter].self)
    }

    public func createFilter(_ body: TodoFilterCreateRequest) async throws(APIError) -> TodoFilter {
        try await client.send(
            .init(.post, "api/v1/filters", body: JSONBody.encode(body)),
            as: TodoFilter.self
        )
    }

    public func updateFilter(
        _ id: String,
        _ body: TodoFilterUpdateRequest
    ) async throws(APIError) -> TodoFilter {
        try await client.send(
            .init(.patch, "api/v1/filters/\(id)", body: JSONBody.encode(body)),
            as: TodoFilter.self
        )
    }

    public func deleteFilter(_ id: String) async throws(APIError) {
        try await client.send(.init(.delete, "api/v1/filters/\(id)"))
    }

    /// Runs the saved query server-side. The client never interprets the filter
    /// syntax itself — that grammar lives in one place, on the server.
    public func runFilter(_ id: String) async throws(APIError) -> [TodoTask] {
        try await client.send(.init(.get, "api/v1/filters/\(id)/tasks"), as: [TodoTask].self)
    }
}
